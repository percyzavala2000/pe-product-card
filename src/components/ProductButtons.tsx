import React, { CSSProperties, HTMLAttributes, useCallback, useContext } from "react";
import styles from '../styles/styles.module.css';
import { ProductContext } from "./ProductCard";

export interface ButtonProps{
    className?:string;
    style?: HTMLAttributes<HTMLElement>| CSSProperties,
   
}
export const ProductButtons = ({className,style}:ButtonProps) => {
    
    const {counter,increaseBy,maxCount}=useContext(ProductContext);

   const isMaxReaChed = useCallback(() => {
        return  !!maxCount && counter===maxCount;
        
      },[counter,maxCount]);
  
      console.log(isMaxReaChed());
      

 
  // true si el count===maxCount
  // false si no lo es
    return (

        <div className={`${styles.buttonsContainer} ${className}`} style={style} >
            <button className={styles.buttonMinus} onClick={() => increaseBy(-1)} >-</button>
            <div className={styles.countLabel}>{counter}</div>
            <button className={`${styles.buttonAdd} ${ isMaxReaChed()&&styles.disabled}`} onClick={() => increaseBy(+ 1)}>
                +
            </button>
        </div>

    )
}