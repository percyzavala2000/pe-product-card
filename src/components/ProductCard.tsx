import React, {createContext} from 'react';
import {useProduct} from '../hooks/useProduct';
import { ProductContextProps, ProductCardProps } from '../interfaces/product.interfaces';
import styles from '../styles/styles.module.css';
/* import { ProductImage } from './ProductImage';
import { ProdcutTitle } from './ProductTitle';
import { ProductButtons } from './ProductButtons'; */


export const ProductContext = createContext({} as ProductContextProps);
const {Provider} = ProductContext;


export const ProductCard = ({children, product,className,style,onChange,value,initialValues} : ProductCardProps) => {
    const {counter, increaseBy,maxCount,isMaxCountReached,reset} = useProduct({onChange, product,value,initialValues});
   
    return (
        <Provider
            value={{
            counter,
            increaseBy,
            product,
            maxCount
        }}>
            <div className={`${styles.productCard} ${className}`} style={style} >
                {children({
                    count: counter,
                    isMaxCountReached,
                    maxCount:initialValues?.maxCount,
                    product,
                    increaseBy,
                    reset

                })}  

                {/* <ProductImage img={product.img} />
            <ProdcutTitle title={product.title} />
        <ProductButtons counter={counter} increaseBy={increaseBy} /> */}

            </div>
        </Provider>
    )
}

/* ProductCard.Title = ProdcutTitle;
ProductCard.Image = ProductImage;
ProductCard.Buttons = ProductButtons; */
