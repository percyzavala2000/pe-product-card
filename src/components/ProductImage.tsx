import React, { CSSProperties, HTMLAttributes, useContext } from "react";
import { ProductContext } from "./ProductCard";
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export interface ImageProps{
    img?:string;
    className?:string;
    style?: HTMLAttributes<HTMLElement>| CSSProperties
}

export const ProductImage = ({img,className,style}:ImageProps) => {
    const {product} = useContext(ProductContext);
    let imgToShow:string;
    if (img) {
        imgToShow = img;
    } else if (product.img) {
        imgToShow = product.img;
    } else {
        imgToShow = noImage;
    }

    
    return (

    <img 
    className={`${styles.productImg} ${className}`}
    style={style}
    src={imgToShow}
    alt="Product" 
    />
    )
};