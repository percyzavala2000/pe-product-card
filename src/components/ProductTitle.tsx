import React, { CSSProperties, HTMLAttributes, useContext } from "react";
import styles from '../styles/styles.module.css';
import { ProductContext } from "./ProductCard";

export interface TitleProps {title?: string,className?:string,style?: HTMLAttributes<HTMLElement>| CSSProperties}


export const ProdcutTitle = ({title,className,style}:TitleProps ) => {

    const {product} = useContext(ProductContext);
    
    
    
    let titleToShow:string;
    if (title) {
        titleToShow = title;
    } else if (product.title) {
        titleToShow = product.title;
    } else {
        titleToShow = 'No Title';
    }

    return (
        <span className={`${styles.productDescription} ${className}`} style={style} >{titleToShow}</span>
    )

};