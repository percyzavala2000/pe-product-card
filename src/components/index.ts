
import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProdcutTitle } from './ProductTitle';
import { ProductCardHOCProps } from '../interfaces/product.interfaces';

export { ProductButtons } from './ProductButtons';
export { ProdcutTitle } from './ProductTitle';
export { ProductImage } from './ProductImage';


export const ProductCard:ProductCardHOCProps=Object.assign(ProductCardHOC,{
    Title:ProdcutTitle,
    Image:ProductImage,
    Buttons:ProductButtons,

});


export default ProductCard;