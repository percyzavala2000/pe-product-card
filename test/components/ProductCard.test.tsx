import  React, { Fragment } from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components/ProductCard';
import { product1 } from '../data/productos';

describe('pruebas en <ProductCard/>', () => {

test('debe de mostrar el componente correctamente', () => {

    const wrapper=renderer.create(
        <ProductCard product={product1} >
            {
                ()=>(
                    <h1>Product card</h1>
                )
            }
        </ProductCard>
    )
    expect(wrapper.toJSON()).toMatchSnapshot();


});
test('debe de incremetar el contador',()=>{
    const wrapper=renderer.create(
        <ProductCard product={product1} >
            {
                ({count,increaseBy})=>(
                    <Fragment>
                    <h1>Product card</h1>
                    <span>{count}</span>
                    <button onClick={()=>increaseBy(1)}>2+</button>
                    </Fragment>
                )
            }
        </ProductCard>
    );
    let tree:any = wrapper.toJSON();
  
    expect(wrapper.toJSON()).toMatchSnapshot();
    
    renderer.act(()=>{
        tree.children[2].props.onClick();

    })
    tree = wrapper.toJSON();
    expect(tree.children[1].children[0]).toBe('1');
    


});

});




