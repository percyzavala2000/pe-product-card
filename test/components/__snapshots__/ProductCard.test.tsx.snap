// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <ProductCard/> debe de incremetar el contador 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    2+
  </button>
</div>
`;

exports[`pruebas en <ProductCard/> debe de mostrar el componente correctamente 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product card
  </h1>
</div>
`;
